src/main/java/io/supertokens/webserver/api/passwordless/CreateCodeAPI.java

```diff
@@ -0,0 +1,118 @@
+ */
+
+package io.supertokens.webserver.api.passwordless;
+
+import com.google.gson.JsonObject;
+import io.supertokens.Main;
+import io.supertokens.config.Config;
+import io.supertokens.passwordless.Passwordless;
+import io.supertokens.passwordless.Passwordless.CreateCodeResponse;
+import io.supertokens.passwordless.exceptions.RestartFlowException;
+import io.supertokens.pluginInterface.RECIPE_ID;
+import io.supertokens.pluginInterface.exceptions.StorageQueryException;
+import io.supertokens.pluginInterface.passwordless.exception.DuplicateLinkCodeHashException;
+import io.supertokens.utils.Utils;
+import io.supertokens.webserver.InputParser;
+import io.supertokens.webserver.WebserverAPI;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+
+public class CreateCodeAPI extends WebserverAPI {
+
+    private static final long serialVersionUID = -4641988458637882374L;
+
+    public CreateCodeAPI(Main main) {
+        super(main, RECIPE_ID.PASSWORDLESS.toString());
+    }
+
+    @Override
+    public String getPath() {
+        return "/recipe/signinup/code";
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException, ServletException {
+        // Logic based on: https://app.code2flow.com/e3yKIdE25SXE
+        JsonObject input = InputParser.parseJsonObjectOrThrowError(req);
+
+        String email = null;
+        String phoneNumber = null;
+        String deviceId = null;
+        if (input.has("email")) {
+            if (input.has("phoneNumber") || input.has("deviceId")) {
+                throw new ServletException(
+                        new BadRequestException("Please provide exactly one of email, phoneNumber or deviceId"));
+            }
+            email = Utils.normaliseEmail(InputParser.parseStringOrThrowError(input, "email", false));
+        } else if (input.has("phoneNumber")) {
+            if (input.has("email") || input.has("deviceId")) {
+                throw new ServletException(
+                        new BadRequestException("Please provide exactly one of email, phoneNumber or deviceId"));
+            }
+            phoneNumber = InputParser.parseStringOrThrowError(input, "phoneNumber", false);
+        } else if (input.has("deviceId")) {
+            if (input.has("email") || input.has("phoneNumber")) {
+                throw new ServletException(
+                        new BadRequestException("Please provide exactly one of email, phoneNumber or deviceId"));
+            }
+            deviceId = InputParser.parseStringOrThrowError(input, "deviceId", false);
+        } else {
+            throw new ServletException(
+                    new BadRequestException("Please provide exactly one of email, phoneNumber or deviceId"));
+        }
```

Comment:
This logic can be made simpler by incrementing a counter each time `email`, `phoneNumber` or `deviceId` are `!= null`. And then throwing this error if the value of that counter is `!= 1` at the end of it (assuming the counter starts from 0)